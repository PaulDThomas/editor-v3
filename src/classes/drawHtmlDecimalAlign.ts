import { EditorV3BlockClass, EditorV3ContentProps, EditorV3RenderProps } from "./interface";

export const drawHtmlDecimalAlign = (
  renderProps: EditorV3RenderProps,
  contentProps: EditorV3ContentProps,
  lhsContent: EditorV3BlockClass[],
  rhsContent: EditorV3BlockClass[],
) => {
  if (renderProps.currentEl) {
    const prePoint = document.createElement("span");
    prePoint.className = "aiev3-span-point lhs";
    renderProps.currentEl.append(prePoint);
    if (lhsContent.length)
      lhsContent.forEach((tb) =>
        prePoint.append(
          tb.toHtml(
            { ...renderProps, currentEl: prePoint },
            tb.style ? contentProps.styles?.[tb.style] : undefined,
          ),
        ),
      );
    else prePoint.textContent = "\u200b";

    // Set up space after (excluding) decimal
    const postPoint = document.createElement("span");
    postPoint.className = "aiev3-span-point rhs";
    renderProps.currentEl.append(postPoint);
    if (rhsContent.length)
      rhsContent.forEach((tb) => {
        postPoint.append(
          tb.toHtml(
            { ...renderProps, currentEl: postPoint },
            tb.style ? contentProps.styles?.[tb.style] : undefined,
          ),
        );
      });
    else postPoint.textContent = "\u200b";

    // Set grid column widths
    renderProps.currentEl.style.gridTemplateColumns = `${contentProps.decimalAlignPercent}% ${100 - contentProps.decimalAlignPercent}%`;
  }
};
