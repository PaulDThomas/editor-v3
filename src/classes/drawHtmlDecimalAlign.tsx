export const drawHtmlDecimalAlign = (
  line: HTMLDivElement,
  fullText: string,
  decimalAlignPercent: number,
) => {
  const prePoint = document.createElement('span');
  prePoint.className = 'aiev3-span aiev3-span-point lhs';
  prePoint.style.right = `${100 - decimalAlignPercent}%`;
  prePoint.style.minWidth = `${decimalAlignPercent}%`;
  line.append(prePoint);

  // Set up space after (excluding) decimal
  const postPoint = document.createElement('span');
  postPoint.className = 'aiev3-span aiev3-span-point rhs';
  postPoint.style.left = `${decimalAlignPercent}%`;
  postPoint.style.minWidth = `${100 - decimalAlignPercent}%`;
  line.append(postPoint);

  // Add spans with text if there is no decimal
  const decimalPosition = fullText.match(/\./)?.index;
  if (decimalPosition !== undefined) {
    prePoint.textContent = fullText.slice(0, decimalPosition + 1).replace(/ /g, '\u00A0');
    postPoint.textContent = fullText.slice(decimalPosition + 1).replace(/ /g, '\u00A0');
  } else {
    prePoint.textContent = fullText !== '' ? fullText.replace(/ /g, '\u00A0') : '\u200b';
    postPoint.textContent = '\u200b';
  }
};
