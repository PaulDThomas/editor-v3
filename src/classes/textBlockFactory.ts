import {
  EditorV3AtBlock,
  IEditorV3AtBlock,
  IEditorV3AtBlockOptionalParams,
} from "./EditorV3AtBlock";
import {
  EditorV3TextBlock,
  IEditorV3TextBlock,
  IEditorV3TextBlockOptionalParams,
} from "./EditorV3TextBlock";

/**
 *  Factory function to direct text block creation
 * @param arg Block data/element
 * @param forcedParams Parameters to force
 * @returns EditorV3TextBlock | EditorV3AtBlock
 */
export const textBlockFactory = (
  arg:
    | HTMLSpanElement
    | Text
    | EditorV3TextBlock
    | DocumentFragment
    | IEditorV3TextBlock
    | IEditorV3AtBlock
    | string,
  forcedParams?: IEditorV3TextBlockOptionalParams | IEditorV3AtBlockOptionalParams,
): EditorV3TextBlock | EditorV3AtBlock => {
  // Span element
  if (arg instanceof HTMLSpanElement) {
    return arg.dataset.type === "at" || forcedParams?.type === "at"
      ? new EditorV3AtBlock(arg, forcedParams)
      : new EditorV3TextBlock(arg, forcedParams);
  }

  // Document Fragment element
  else if (arg instanceof DocumentFragment) {
    return arg.childNodes.length > 0 &&
      arg.childNodes[0] instanceof HTMLSpanElement &&
      (arg.childNodes[0].dataset.type === "at" || forcedParams?.type === "at")
      ? new EditorV3AtBlock(arg, forcedParams)
      : new EditorV3TextBlock(arg, forcedParams);
  }
  // Text node
  else if (arg instanceof Text) {
    return new EditorV3TextBlock(
      { text: (arg.textContent ?? "").replaceAll("\u00A0", " ") },
      forcedParams,
    );
  }
  // Text
  else if (typeof arg === "string") {
    try {
      const jsonInput = JSON.parse(arg);
      return jsonInput.type === "at" || forcedParams?.type === "at"
        ? new EditorV3AtBlock(jsonInput, forcedParams)
        : typeof jsonInput !== "object"
          ? new EditorV3TextBlock({ text: `${jsonInput}` }, forcedParams)
          : new EditorV3TextBlock(jsonInput, forcedParams);
    } catch {
      return new EditorV3TextBlock({ text: arg }, forcedParams);
    }
  }
  // Must be object
  else {
    return arg.type === "at" || forcedParams?.type === "at"
      ? new EditorV3AtBlock(arg, forcedParams)
      : new EditorV3TextBlock(arg, forcedParams);
  }
};
