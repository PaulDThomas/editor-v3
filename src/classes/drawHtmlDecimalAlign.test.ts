/* eslint-disable quotes */
import { drawHtmlDecimalAlign } from "./drawHtmlDecimalAlign";
import { textBlockFactory } from "./textBlockFactory";

describe("Test draw decimal align function", () => {
  test("Draw with point", async () => {
    const div = document.createElement("div");
    drawHtmlDecimalAlign(
      { currentEl: div },
      40,
      [textBlockFactory("12.")],
      [textBlockFactory("34")],
    );
    expect(div.outerHTML).toEqual(
      '<div style="grid-template-columns: 40% 60%;">' +
        '<span class="aiev3-span-point lhs"><span class="aiev3-tb">12.\u200c</span></span>' +
        '<span class="aiev3-span-point rhs"><span class="aiev3-tb">34\u200c</span></span>' +
        "</div>",
    );
  });
  test("Draw without right", async () => {
    const div = document.createElement("div");
    drawHtmlDecimalAlign({ currentEl: div }, 40, [textBlockFactory("1234")], []);
    expect(div.outerHTML).toEqual(
      '<div style="grid-template-columns: 40% 60%;">' +
        '<span class="aiev3-span-point lhs"><span class="aiev3-tb">1234\u200c</span></span>' +
        '<span class="aiev3-span-point rhs">\u2009</span>' +
        "</div>",
    );
  });
  test("Draw without left", async () => {
    const div = document.createElement("div");
    drawHtmlDecimalAlign({ currentEl: div }, 75, [], [textBlockFactory("34")]);
    expect(div.outerHTML).toEqual(
      '<div style="grid-template-columns: 75% 25%;">' +
        '<span class="aiev3-span-point lhs">\u2009</span>' +
        '<span class="aiev3-span-point rhs"><span class="aiev3-tb">34\u200c</span></span>' +
        "</div>",
    );
  });
  test("Draw empty", async () => {
    const div = document.createElement("div");
    drawHtmlDecimalAlign({ currentEl: div }, 35, [], []);
    expect(div.outerHTML).toEqual(
      '<div style="grid-template-columns: 35% 65%;">' +
        '<span class="aiev3-span-point lhs">\u2009</span>' +
        '<span class="aiev3-span-point rhs">\u2009</span>' +
        "</div>",
    );
  });
});
