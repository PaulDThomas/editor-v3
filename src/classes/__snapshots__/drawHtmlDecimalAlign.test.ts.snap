// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test draw decimal align function Draw with point 1`] = `"<div style="grid-template-columns: 40% 60%;"><span class="aiev3-span-point lhs"><span class="aiev3-tb" data-line-start-position="0">12.</span></span><span class="aiev3-span-point rhs"><span class="aiev3-tb" data-line-start-position="0">34</span></span></div>"`;

exports[`Test draw decimal align function Draw without left 1`] = `"<div style="grid-template-columns: 75% 25%;"><span class="aiev3-span-point lhs">​</span><span class="aiev3-span-point rhs"><span class="aiev3-tb" data-line-start-position="0">34</span></span></div>"`;

exports[`Test draw decimal align function Draw without right 1`] = `"<div style="grid-template-columns: 40% 60%;"><span class="aiev3-span-point lhs"><span class="aiev3-tb" data-line-start-position="0">1234</span></span><span class="aiev3-span-point rhs">​</span></div>"`;
